pkg/collector/container/container.go

```diff
@@ -228,167 +225,159 @@ func (c *Collector) collect(ch chan<- prometheus.Metric) error {
-			)
-			ch <- prometheus.MustNewConstMetric(
-				c.usagePrivateWorkingSetBytes,
-				prometheus.GaugeValue,
-				float64(cstats.Memory.UsagePrivateWorkingSetBytes),
-				containerIdWithPrefix,
-			)
-			ch <- prometheus.MustNewConstMetric(
-				c.runtimeTotal,
-				prometheus.CounterValue,
-				float64(cstats.Processor.TotalRuntime100ns)*perflib.TicksToSecondScaleFactor,
-				containerIdWithPrefix,
-			)
-			ch <- prometheus.MustNewConstMetric(
-				c.runtimeUser,
-				prometheus.CounterValue,
-				float64(cstats.Processor.RuntimeUser100ns)*perflib.TicksToSecondScaleFactor,
-				containerIdWithPrefix,
-			)
-			ch <- prometheus.MustNewConstMetric(
-				c.runtimeKernel,
-				prometheus.CounterValue,
-				float64(cstats.Processor.RuntimeKernel100ns)*perflib.TicksToSecondScaleFactor,
-				containerIdWithPrefix,
-			)
-			ch <- prometheus.MustNewConstMetric(
-				c.readCountNormalized,
-				prometheus.CounterValue,
-				float64(cstats.Storage.ReadCountNormalized),
-				containerIdWithPrefix,
-			)
-			ch <- prometheus.MustNewConstMetric(
-				c.readSizeBytes,
-				prometheus.CounterValue,
-				float64(cstats.Storage.ReadSizeBytes),
-				containerIdWithPrefix,
-			)
-			ch <- prometheus.MustNewConstMetric(
-				c.writeCountNormalized,
-				prometheus.CounterValue,
-				float64(cstats.Storage.WriteCountNormalized),
-				containerIdWithPrefix,
-			)
-			ch <- prometheus.MustNewConstMetric(
-				c.writeSizeBytes,
-				prometheus.CounterValue,
-				float64(cstats.Storage.WriteSizeBytes),
-				containerIdWithPrefix,
-			)
-		}()
+		err := c.collectContainer(ch, containerDetails)
+
+		if hcsshim.IsNotExist(err) {
+			_ = level.Debug(c.logger).Log("msg", "err in fetching container statistics", "containerId", containerDetails.ID, "err", err)
+		} else {
+			_ = level.Error(c.logger).Log("msg", "err in fetching container statistics", "containerId", containerDetails.ID, "err", err)
+			hasErrors = true
+		}
 	}
 
-	hnsEndpoints, err := hcsshim.HNSListEndpointRequest()
-	if err != nil {
-		_ = level.Warn(c.logger).Log("msg", "Failed to collect network stats for containers")
-		return err
+	if hasErrors {
+		return errors.New("errors while fetching container statistics")
 	}
 
-	if len(hnsEndpoints) == 0 {
-		_ = level.Info(c.logger).Log("msg", "No network stats for containers to collect")
-		return nil
+	return nil
+}
+
+func (c *Collector) collectContainer(ch chan<- prometheus.Metric, containerDetails hcsshim.ContainerProperties) error {
+	container, err := hcsshim.OpenContainer(containerDetails.ID)
+	if err != nil {
+		return fmt.Errorf("error in opening container: %w", err)
 	}
 
-	for _, endpoint := range hnsEndpoints {
-		endpointStats, err := hcsshim.GetHNSEndpointStats(endpoint.Id)
-		if err != nil {
-			_ = level.Warn(c.logger).Log("msg", "Failed to collect network stats for interface "+endpoint.Id, "err", err)
-			continue
+	defer func() {
+		if container == nil {
+			return
 		}
 
-		for _, containerId := range endpoint.SharedContainers {
```

Comment:
you've dropped this,  It is tempting to use the `.network` to collect the stats but you can't do that with containerd's hns v2.  `.network` only works on v1 hns apis.  
see https://github.com/prometheus-community/windows_exporter/commit/9adefdceeb6a6467bda2222f124e72f6a8081e36 and https://github.com/prometheus-community/windows_exporter/pull/1218

Reply:
Missing knowledge and missing comments, thanks for the hints!

Reply:
I have re-add this, thanks a lot!